tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/subbuilder/SetBuilder.scala

```diff
@@ -0,0 +1,117 @@
+package at.forsyte.apalache.tla.typecmp.subbuilder
+
+import at.forsyte.apalache.tla.lir._
+import at.forsyte.apalache.tla.lir.oper.TlaOper
+import at.forsyte.apalache.tla.typecmp.BuilderUtil._
+import at.forsyte.apalache.tla.typecmp.{BuilderWrapper, InternalState, NameWrapper, ValWrapper}
+import at.forsyte.apalache.tla.typecmp.raw.RawSetBuilder
+import scalaz._
+import scalaz.Scalaz._
+
+/**
+ * Builder for TlaSetOper expressions.
+ *
+ * @author
+ *   Jure Kukovec
+ */
+trait SetBuilder extends RawSetBuilder {
+
+  /** {x1, ..., xn}, must have >0 args. */
+  def enumSet(argW: BuilderWrapper, argsW: BuilderWrapper*): BuilderWrapper = for {
+    argEx <- argW
+    argsExs <- buildSeq(argsW)
+  } yield _enumSet(argEx, argsExs: _*)
+
+  /** {}: Set(elemType) */
+  def emptySet(elemType: TlaType1): BuilderWrapper = _emptySet(elemType).point[InternalState]
+
+  /** elem \in set */
+  def in(elemW: BuilderWrapper, setW: BuilderWrapper): BuilderWrapper =
+    binaryFromRaw(elemW, setW)(_in)
+
+  /** elem \notin set */
+  def notin(elemW: BuilderWrapper, setW: BuilderWrapper): BuilderWrapper =
+    binaryFromRaw(elemW, setW)(_notin)
+
+  /** left \cap right, left \intersect right */
+  def cap(leftW: BuilderWrapper, rightW: BuilderWrapper): BuilderWrapper =
+    binaryFromRaw(leftW, rightW)(_cap)
+
+  /** left \cup right, left `\`union right */
+  def cup(leftW: BuilderWrapper, rightW: BuilderWrapper): BuilderWrapper =
+    binaryFromRaw(leftW, rightW)(_cup)
+
+  /** UNION set */
+  def union(setW: BuilderWrapper): BuilderWrapper = setW.map(_union)
+
+  /** { x \in set: p } */
+  def filter(xW: NameWrapper, setW: BuilderWrapper, pW: BuilderWrapper): BuilderWrapper = for {
+    x <- xW
+    set <- setW
+    p <- pW
+    _ <- markAsBound(x)
+  } yield _filter(x, set, p)
+
+  /** { mapExpr: x1 \in set1 , ..., xN \in setN }, must have at least 1 var-set pair */
+  def map(
+      eW: BuilderWrapper,
+      x1W: NameWrapper,
+      set1W: BuilderWrapper,
+      pairsW: BuilderWrapper*): BuilderWrapper = for {
```

Comment:
Same here. I would prefer a uniform interface. Perhaps, a sequence of pairs? Scala's `->` can be very nicely applied here.

Reply:
I can make a `BuilderWrapper*` variant and a `(A,B)*` variant, like I do with record sets.